# IMPORTACION DE LIBRERIAS Y DATOS

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import FunctionTransformer

In [ ]:
from google.colab import drive
drive.mount('/content/Drive')

Drive already mounted at /content/Drive; to attempt to forcibly remount, call drive.mount("/content/Drive", force_remount=True).


In [ ]:
df_sales = pd.read_csv('/content/Drive/MyDrive/datasets_TFM/sales.csv')
df_product_desc = pd.read_csv('/content/Drive/MyDrive/datasets_TFM/product_description.csv')
df_customdemo = pd.read_csv('/content/Drive/MyDrive/datasets_TFM/customer_sociodemographics.csv')
df_custom_prod = pd.read_csv('/content/Drive/MyDrive/datasets_TFM/customer_products.csv')
df_customcommact = pd.read_csv('/content/Drive/MyDrive/datasets_TFM/customer_commercial_activity.csv')

In [ ]:
dfs = {
    "df_sales": df_sales,
    "df_product_desc": df_product_desc,
    "df_customdemo": df_customdemo,
    "df_custom_prod": df_custom_prod,
    "df_customcommact": df_customcommact
}


# DATA UNDERSTANDING

In [ ]:
def check_data_understanding(df, nombre_dataset):
    print(f"{'='*30}")
    print(f"ANÁLISIS DE: {nombre_dataset.upper()}")
    print(f"{'='*30}")

    # 1. Dimensiones y Duplicados
    print(f" Filas: {df.shape[0]} | Columnas: {df.shape[1]}")
    print(f" Filas duplicadas: {df.duplicated().sum()}")

    # Valores unicos
    print(f'\n Valores Unicos:{df.nunique()}')

    # 2. Tipos de datos y Nulos
    print("\n Información de Columnas y Nulos:")
    info_df = pd.DataFrame({
        'Tipo': df.dtypes,
        'Nulos': df.isnull().sum(),
        '% Nulos': (df.isnull().sum() / len(df) * 100).round(2)
    })
    print(info_df)

    # 3. Análisis de variables numéricas (para ver outliers o rangos)
    if not df.select_dtypes(include=[np.number]).empty:
        print("\n Resumen Estadístico (Numéricas):")
        print(df.describe().T.round(2))

    # 4. Muestra de datos
    print("\n Primeras 3 filas:")
    display(df.head(3))
    print("\n")

# Una vez tengas cargados tus dataframes, solo tienes que llamarla así:

check_data_understanding(df_sales, "Ventas")
check_data_understanding(df_product_desc, "Descripción Productos")
check_data_understanding(df_customdemo, "Sociodemográficos")
check_data_understanding(df_custom_prod, "Productos Cliente")
check_data_understanding(df_customcommact, "Actividad Comercial")

ANÁLISIS DE: VENTAS
 Filas: 240773 | Columnas: 6
 Filas duplicadas: 0

 Valores Unicos:Unnamed: 0    240773
pk_sale       240773
cid           152754
month_sale        16
product_ID        13
net_margin     25451
dtype: int64

 Información de Columnas y Nulos:
               Tipo  Nulos  % Nulos
Unnamed: 0    int64      0      0.0
pk_sale       int64      0      0.0
cid           int64      0      0.0
month_sale   object      0      0.0
product_ID    int64      0      0.0
net_margin  float64      0      0.0

 Resumen Estadístico (Numéricas):
               count        mean        std      min        25%        50%  \
Unnamed: 0  240773.0   120386.00   69505.32      0.0    60193.0   120386.0   
pk_sale     240773.0   127052.00   69505.32   6666.0    66859.0   127052.0   
cid         240773.0  1363933.72  175172.07  15891.0  1301012.0  1409253.0   
product_ID  240773.0     3945.41    1632.82   1119.0     2312.0     4657.0   
net_margin  240773.0      606.88    1735.81     25.0       61.

,Unnamed: 0,pk_sale,cid,month_sale,product_ID,net_margin
0,0,6666,33620,2018-05-01,2335,952.9
1,1,6667,35063,2018-06-01,2335,1625.2
2,2,6668,37299,2018-02-01,2335,1279.7




ANÁLISIS DE: DESCRIPCIÓN PRODUCTOS
 Filas: 13 | Columnas: 4
 Filas duplicadas: 0

 Valores Unicos:Unnamed: 0        13
pk_product_ID     13
product_desc      13
family_product     5
dtype: int64

 Información de Columnas y Nulos:
                  Tipo  Nulos  % Nulos
Unnamed: 0       int64      0      0.0
pk_product_ID    int64      0      0.0
product_desc    object      0      0.0
family_product  object      0      0.0

 Resumen Estadístico (Numéricas):
               count     mean      std     min     25%     50%     75%     max
Unnamed: 0      13.0     6.00     3.89     0.0     3.0     6.0     9.0    12.0
pk_product_ID   13.0  3614.54  2564.25  1119.0  2235.0  2336.0  4033.0  9001.0

 Primeras 3 filas:


,Unnamed: 0,pk_product_ID,product_desc,family_product
0,0,4657,em_acount,account
1,1,3819,debit_card,payment_card
2,2,1364,pension_plan,pension_plan




ANÁLISIS DE: SOCIODEMOGRÁFICOS
 Filas: 5962924 | Columnas: 9
 Filas duplicadas: 0

 Valores Unicos:Unnamed: 0      5962924
pk_cid           456373
pk_partition         17
country_id           41
region_code          52
gender                2
age                 104
deceased              2
salary           257654
dtype: int64

 Información de Columnas y Nulos:
                 Tipo    Nulos  % Nulos
Unnamed: 0      int64        0     0.00
pk_cid          int64        0     0.00
pk_partition   object        0     0.00
country_id     object        0     0.00
region_code   float64     2264     0.04
gender         object       25     0.00
age             int64        0     0.00
deceased       object        0     0.00
salary        float64  1541104    25.84

 Resumen Estadístico (Numéricas):
                 count        mean         std       min         25%  \
Unnamed: 0   5962924.0  7155425.01  3920784.28      0.00  4240385.75   
pk_cid       5962924.0  1234929.80   162302.04  15891.00

,Unnamed: 0,pk_cid,pk_partition,country_id,region_code,gender,age,deceased,salary
0,0,1375586,2018-01,ES,29.0,H,35,N,87218.10
1,1,1050611,2018-01,ES,13.0,V,23,N,35548.74
2,2,1050612,2018-01,ES,13.0,V,23,N,122179.11




ANÁLISIS DE: PRODUCTOS CLIENTE
 Filas: 5962924 | Columnas: 18
 Filas duplicadas: 0

 Valores Unicos:Unnamed: 0            5962924
pk_cid                 456373
pk_partition               17
short_term_deposit          2
loans                       2
mortgage                    2
funds                       2
securities                  2
long_term_deposit           2
em_account_pp               1
credit_card                 2
payroll                     2
pension_plan                2
payroll_account             2
emc_account                 2
debit_card                  2
em_account_p                2
em_acount                   2
dtype: int64

 Información de Columnas y Nulos:
                       Tipo  Nulos  % Nulos
Unnamed: 0            int64      0      0.0
pk_cid                int64      0      0.0
pk_partition         object      0      0.0
short_term_deposit    int64      0      0.0
loans                 int64      0      0.0
mortgage              int64      0      0.0
fu

,Unnamed: 0,pk_cid,pk_partition,short_term_deposit,loans,mortgage,funds,securities,long_term_deposit,em_account_pp,credit_card,payroll,pension_plan,payroll_account,emc_account,debit_card,em_account_p,em_acount
0,0,1375586,2018-01,0,0,0,0,0,0,0,0,0.0,0.0,0,0,0,0,1
1,1,1050611,2018-01,0,0,0,0,0,0,0,0,0.0,0.0,0,0,0,0,1
2,2,1050612,2018-01,0,0,0,0,0,0,0,0,0.0,0.0,0,0,0,0,1




ANÁLISIS DE: ACTIVIDAD COMERCIAL
 Filas: 5962924 | Columnas: 7
 Filas duplicadas: 0

 Valores Unicos:Unnamed: 0         5962924
pk_cid              456373
pk_partition            17
entry_date              53
entry_channel           68
active_customer          2
segment                  3
dtype: int64

 Información de Columnas y Nulos:
                    Tipo   Nulos  % Nulos
Unnamed: 0         int64       0     0.00
pk_cid             int64       0     0.00
pk_partition      object       0     0.00
entry_date        object       0     0.00
entry_channel     object  133033     2.23
active_customer  float64       0     0.00
segment           object  133944     2.25

 Resumen Estadístico (Numéricas):
                     count        mean         std      min         25%  \
Unnamed: 0       5962924.0  7155425.01  3920784.28      0.0  4240385.75   
pk_cid           5962924.0  1234929.80   162302.04  15891.0  1112532.00   
active_customer  5962924.0        0.40        0.49      0.0     

,Unnamed: 0,pk_cid,pk_partition,entry_date,entry_channel,active_customer,segment
0,0,1375586,2018-01,2018-01,KHL,1.0,02 - PARTICULARES
1,1,1050611,2018-01,2015-08,KHE,0.0,03 - UNIVERSITARIO
2,2,1050612,2018-01,2015-08,KHE,0.0,03 - UNIVERSITARIO


### RESUMEN DE LA INFORMACION DE LOS DATOS:

1. Clientes
Lo más raro que ha salido es la edad: hay clientes con 2 años y otros con más de 100. Lo de los niños serán cuentas de ahorro de los padres, pero a nosotros nos estorban porque un bebé no te va a contratar una tarjeta o un préstamo. Si no los quitamos, nos van a romper todas las estadísticas y las medias de edad no van a tener sentido.

2. Salarios
En los sueldos hay un caos total. Primero, hay un 25% de nulos (gente sin salario), lo que nos deja un poco a ciegas. Y segundo, hay gente ganando 28 millones de euros, mientras que la mayoría está en 88 mil euros. Ese outlier (28 millones de euros) hace que la media suba artificialmente, así que habrá que limitar esos valores máximos para que los datos sean más realistas.

3. Entry_channel
En el canal de entrada (entry_channel) nos encontramos codigos, dichos codigos Son las siglas que usa el banco para saber por dónde llegó el cliente (web, oficina, etc.). Como están muy fragmentados y hay muchísimos, habrá que agruparlos para entender de dónde vienen los clientes que de verdad dejan dinero.

4. datos y fechas de texto
Por último, tenemos columnas como Unnamed: 0 que son puro relleno y hay que borrarlas porque no sirven para nada. Además, las fechas de las particiones (pk_partition) están guardadas como si fueran texto. Si no las pasamos a formato fecha, no podremos hacer gráficas para ver si el banco está creciendo o perdiendo clientes mes a mes.

Tras el análisis exploratorio inicial, se han identificado variables críticas como entry_channel, segment y las familias de productos. Debido a su alta cardinalidad (muchos valores distintos) y su importancia para el negocio, se realizará un análisis profundo de sus frecuencias para definir la mejor estrategia de agrupación. Una vez establecida la estrategia, se procederá a la fase de Data Preparation.

### Enty_channel = Canal de captación del cliente

In [ ]:
pd.set_option('display.max_rows', 100)
display(pd.Series(df_customcommact['entry_channel'].unique()))

,0
0,KHL
1,KHE
2,KHD
3,KFA
4,KFC
5,KAT
6,KAZ
7,RED
8,KHC
9,KHK


In [ ]:
display(pd.Series(df_customcommact['entry_channel']).value_counts())

,count
entry_channel,
KHE,3113947
KFC,890620
KHQ,590280
KAT,416084
KHK,230197
KHM,176591
KHN,108434
KFA,79020
KHD,75899


### Segment = Tipo de cliente

In [ ]:
pd.Series(df_customcommact['segment'].unique())

,0
0,02 - PARTICULARES
1,03 - UNIVERSITARIO
2,01 - TOP
3,NaN


In [ ]:
df_customcommact['segment'].isnull().value_counts()

,count
segment,
False,5828980
True,133944


In [ ]:
df_customcommact['segment'].value_counts()

,count
segment,
03 - UNIVERSITARIO,3900166
02 - PARTICULARES,1830875
01 - TOP,97939


## Family_Product Grupo al que pertenece.

In [ ]:
pd.Series(df_product_desc['family_product'].unique())

,0
0,account
1,payment_card
2,pension_plan
3,investment
4,loan


se ha realizado una inspección de la variable deceased (indicador de fallecimiento) al presentarse en formato de texto. Con el objetivo de evaluar su utilidad para el modelo, se ha procedido a contabilizar sus frecuencias. Este análisis permitirá determinar si la variable posee suficiente variabilidad o si, por el contrario, debe descartarse por carecer de poder predictivo

In [ ]:
df_customdemo['deceased'].value_counts()

,count
deceased,
N,5961849
S,1075


COUNTRY ID

In [ ]:
df_customdemo['country_id'].value_counts()

,count
country_id,
ES,5960672
GB,441
FR,225
DE,199
US,195
CH,194
BR,87
BE,81
VE,79


In [ ]:
counts = df_customdemo['country_id'].value_counts()

# 2. Calculamos el porcentaje relativo (normalizamos los valores)
percentages = df_customdemo['country_id'].value_counts(normalize=True) * 100
percentages

,proportion
country_id,
ES,99.962233
GB,0.007396
FR,0.003773
DE,0.003337
US,0.003270
CH,0.003253
BR,0.001459
BE,0.001358
VE,0.001325


Se ha llevado a cabo un análisis de representatividad geográfica sobre la variable country_id con el fin de evaluar si la ubicación del cliente influye en la contratación de productos. Los resultados muestran que el 99.96% de los registros pertenecen a España (ES), lo que evidencia una concentración extrema de la muestra en un único mercado. Debido a esta práctica ausencia de variabilidad, la variable no ofrece capacidad de segmentación para el algoritmo, por lo que se ha decidido excluirla del modelo para evitar el procesamiento de datos redundantes.

In [ ]:
df_customdemo['region_code'].value_counts()

,count
region_code,
28.0,1185757
8.0,600362
46.0,358142
30.0,296856
41.0,290447
15.0,254891
29.0,187911
36.0,175850
3.0,171358


## Limpieza de valores nulos

Tras el análisis exploratorio realizado en la fase de Data Understanding, se ha detectado la presencia de valores nulos en variables clave, por lo que se ha diseñado una estrategia de imputación diferenciada según su naturaleza para evitar la pérdida de registros.

Para las variables categóricas como region_code, entry_channel y segment, los valores faltantes se agruparán bajo la etiqueta 'Other', evitando así sesgar el modelo hacia categorías específicas; por su parte, en las variables gender, payroll y pension_plan se aplicará una imputación por moda al ser el valor más frecuente y representativo de la muestra.

Finalmente, para la variable numérica salary, se optará por la mediana como medida de tendencia central, garantizando una imputación robusta frente a los valores extremos detectados y manteniendo la integridad técnica de la columna para el posterior modelado.

In [ ]:
for name, df in dfs.items():
  nulls_count = df.isnull().sum()
  if nulls_count.sum()>0:
    print(f'\n Dataset {name}')
    print(nulls_count[nulls_count>0])


 Dataset df_customdemo
region_code       2264
gender              25
salary         1541104
dtype: int64

 Dataset df_custom_prod
payroll         61
pension_plan    61
dtype: int64

 Dataset df_customcommact
entry_channel    133033
segment          133944
dtype: int64


In [ ]:
def universal_cleaner(df):
    # 1. Definimos los grupos de variables según la estrategia
    to_other = ['region_code', 'entry_channel', 'segment']
    to_moda = ['gender', 'payroll', 'pension_plan']
    to_median = ['salary']

    # 2. Aplicamos la lógica solo si la columna existe en el dataset
    for col in df.columns:
        if col in to_other:
            df[col] = df[col].fillna('Other')

        elif col in to_moda:
            # Solo si hay nulos calculamos la moda para evitar errores
            if df[col].isnull().any():
                df[col] = df[col].fillna(df[col].mode()[0])

        elif col in to_median:
            df[col] = df[col].fillna(df[col].median())

    return df

# 3. LLAMADA A LA FUNCIÓN
df_sales = universal_cleaner(df_sales)
df_product_desc = universal_cleaner(df_product_desc)
df_customdemo = universal_cleaner(df_customdemo)
df_custom_prod = universal_cleaner(df_custom_prod)
df_customcommact = universal_cleaner(df_customcommact)

# 4. COMPROBACIÓN FINAL
print("Comprobación de nulos tras la limpieza:")
print(f"- Salary en customdemo: {df_customdemo['salary'].isnull().sum()}")
print(f"- Entry Channel en customcommact: {df_customcommact['entry_channel'].isnull().sum()}")
print(f"- Gender en customdemo: {df_customdemo['gender'].isnull().sum()}")

Comprobación de nulos tras la limpieza:
- Salary en customdemo: 0
- Entry Channel en customcommact: 0
- Gender en customdemo: 0


In [ ]:
df_customcommact['segment'].value_counts()

,count
segment,
03 - UNIVERSITARIO,3900166
02 - PARTICULARES,1830875
Other,133944
01 - TOP,97939


# Data Preparation

Una vez limpio los nulos se procede a unir los datasets para el data preparation

In [ ]:
df_merge = pd.merge(df_customdemo, df_customcommact, on=['pk_cid', 'pk_partition'], how='left')

In [ ]:
df_merge

,Unnamed: 0_x,pk_cid,pk_partition,country_id,region_code,gender,age,deceased,salary,Unnamed: 0_y,entry_date,entry_channel,active_customer,segment
0,0,1375586,2018-01,ES,29.0,H,35,N,87218.10,0,2018-01,KHL,1.0,02 - PARTICULARES
1,1,1050611,2018-01,ES,13.0,V,23,N,35548.74,1,2015-08,KHE,0.0,03 - UNIVERSITARIO
2,2,1050612,2018-01,ES,13.0,V,23,N,122179.11,2,2015-08,KHE,0.0,03 - UNIVERSITARIO
3,3,1050613,2018-01,ES,50.0,H,22,N,119775.54,3,2015-08,KHD,0.0,03 - UNIVERSITARIO
4,4,1050614,2018-01,ES,50.0,V,23,N,88441.47,4,2015-08,KHE,1.0,03 - UNIVERSITARIO
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5962919,13647304,1166765,2019-05,ES,50.0,V,22,N,43912.17,13647304,2016-08,KHE,0.0,03 - UNIVERSITARIO
5962920,13647305,1166764,2019-05,ES,26.0,V,23,N,23334.99,13647305,2016-08,KHE,0.0,03 - UNIVERSITARIO
5962921,13647306,1166763,2019-05,ES,50.0,H,47,N,88441.47,13647306,2016-08,KHE,1.0,02 - PARTICULARES
5962922,13647307,1166789,2019-05,ES,50.0,H,22,N,199592.82,13647307,2016-08,KHE,0.0,03 - UNIVERSITARIO


In [ ]:
df_merge = pd.merge(df_merge, df_custom_prod, on=['pk_cid', 'pk_partition'], how='left')

In [ ]:
# 1. Limpiamos las columnas "Unnamed" que sobran en ambos datasets
cols_to_drop = [c for c in df_merge.columns if 'Unnamed' in c]
df_merge = df_merge.drop(columns=cols_to_drop)

cols_to_drop_sales = [c for c in df_sales.columns if 'Unnamed' in c]
df_sales = df_sales.drop(columns=cols_to_drop_sales)

# 2. Aseguramos el nombre del id
if 'cid' in df_sales.columns:
    df_sales = df_sales.rename(columns={'cid': 'pk_cid'})

# 3. Hacemos el merge limpio por pk_cid
# df_merge = pd.merge(df_merge, df_sales, on='pk_cid', how='left')



In [ ]:
# Convertimos 'pk_partition' (ej: '2019-02') a un periodo mensual 'YYYY-MM'
# Esto facilita que Pandas lo entienda como tiempo y no como un simple texto
df_merge['month_key'] = pd.to_datetime(df_merge['pk_partition']).dt.to_period('M')

# Para 'month_sale' (ej: '2019-02-01'), al convertirlo a periodo 'M',
# Pandas automáticamente ignora el '-01' y lo alinea con las otras tablas
df_sales['month_key'] = pd.to_datetime(df_sales['month_sale']).dt.to_period('M')

# -------------------------------------------------------------------
# AGREGACIÓN DE VENTAS (EVITAR DUPLICADOS)
# -------------------------------------------------------------------

# Sumamos el margen neto de todas las ventas que un cliente hizo en un mismo mes
sales_monthly = df_sales.groupby(['pk_cid', 'month_key'])['net_margin'].sum().reset_index()

# -------------------------------------------------------------------
# EL MERGE MAESTRO (AÑADIR EL TARGET / RENTABILIDAD)
# -------------------------------------------------------------------

# Unimos con las ventas usando 'left join'.
# Los clientes que NO compraron ese mes tendrán un margen de NaN.
df_merge = pd.merge(df_merge, sales_monthly,
                    on=['pk_cid', 'month_key'],
                    how='left')

# -------------------------------------------------------------------
# LIMPIEZA FINAL
# -------------------------------------------------------------------
# Rellenamos los NaN de 'net_margin' con 0 (lógico: no hubo compra, no hay margen)
df_merge['net_margin'] = df_merge['net_margin'].fillna(0)

# Borramos la columna auxiliar 'month_key' para no ocupar RAM innecesariamente
df_merge.drop(columns=['month_key'], inplace=True, errors='ignore')

print(f"Merge exitoso. Filas de la tabla final: {df_merge.shape[0]}")

Merge exitoso. Filas de la tabla final: 5962924


In [ ]:
df_merge

,pk_cid,pk_partition,country_id,region_code,gender,age,deceased,salary,entry_date,entry_channel,...,em_account_pp,credit_card,payroll,pension_plan,payroll_account,emc_account,debit_card,em_account_p,em_acount,net_margin
0,1375586,2018-01,ES,29.0,H,35,N,87218.10,2018-01,KHL,...,0,0,0.0,0.0,0,0,0,0,1,0.0
1,1050611,2018-01,ES,13.0,V,23,N,35548.74,2015-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
2,1050612,2018-01,ES,13.0,V,23,N,122179.11,2015-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
3,1050613,2018-01,ES,50.0,H,22,N,119775.54,2015-08,KHD,...,0,0,0.0,0.0,0,0,0,0,0,0.0
4,1050614,2018-01,ES,50.0,V,23,N,88441.47,2015-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5962919,1166765,2019-05,ES,50.0,V,22,N,43912.17,2016-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
5962920,1166764,2019-05,ES,26.0,V,23,N,23334.99,2016-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
5962921,1166763,2019-05,ES,50.0,H,47,N,88441.47,2016-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0
5962922,1166789,2019-05,ES,50.0,H,22,N,199592.82,2016-08,KHE,...,0,0,0.0,0.0,0,0,0,0,1,0.0


In [ ]:
df_merge['em_account_pp'].value_counts()

,count
em_account_pp,
0,5962924


In [ ]:
df_merge['deceased'].value_counts()

,count
deceased,
N,5961849
S,1075


Eliminamos las filas de los clientes muertos ya que al momento de la prospension y segmentacion estos clientes no son relevantes. Si se dejan el modelo los podría "elegir" como parte de la campaña. Es mejor eliminarlos y luego eliminar la colummna.

In [ ]:
print(f"Filas antes: {len(df_merge)}")
df_merge = df_merge[df_merge['deceased'] == 'N']
print(f"Filas después: {len(df_merge)}")

Filas antes: 5962924
Filas después: 5961849


### Eliminacion de valores no relevantes para el modelo

Tras un análisis exhaustivo de la relevancia de las variables para la fase de modelado, se ha determinado la eliminación de ciertos atributos basándose en los siguientes criterios técnicos:

em_account_pp (Varianza Cero): Se ha identificado que esta variable presenta un único valor constante en todo el dataset. Al no existir variabilidad, la columna no aporta información predictiva, ya que no permite al modelo distinguir patrones ni establecer diferencias entre los sujetos del estudio.

deceased (Desequilibrio Extremo): Esta variable registra si el cliente ha fallecido. Anteriormente se eliminaron los clientes fallecidos ya que no tiene sentido incluirlos en las futuras campañas. Ahora esta columna solo tiene valores de clientes vivos, por tanto se puede elimnar ya que no agregar valor a los modelos.

En el caso de country_id se decide eliminar tal y como se ha comentado en el data understanding debido a que Los resultados muestran que el 99.96% de los registros pertenecen a España (ES), lo que evidencia una concentración extrema de la muestra en un único mercado.

In [ ]:
cols_to_drop = [
    'em_account_pp', 'deceased', 'country_id', 'Unnamed: 0'
]
df_reduced = df_merge.drop(columns=[c for c in cols_to_drop if c in df_merge.columns])

Una vez consolidado el dataset final (df_reduced), se ha procedido a categorizar las variables en grupos lógicos según su naturaleza y el tratamiento técnico que requieren. Esta segmentación mediante listas permite automatizar la aplicación de estrategias de transformación específicas —como el desglose de fechas, la codificación de variables categóricas o la normalización de valores numéricos— garantizando un preprocesamiento estructurado y eficiente para el modelo de Machine Learning.

In [ ]:

def auto_classify_columns(df):
    """
    Clasifica automáticamente las columnas del dataframe según su contenido
    y tipo de dato
    """

    # 1. FECHAS: Identifica columnas que contienen 'date', 'partition' o 'month' en el nombre
    list_dates = [col for col in df.columns if any(x in col.lower() for x in ['date', 'partition', 'month_sale'])]

    # 2. BINARIAS/PRODUCTOS: Columnas con solo 2 valores únicos
    # Excluimos las que ya son fechas
    list_binary = [col for col in df.columns if df[col].nunique() == 2 and col not in list_dates]

    # 3. NUMÉRICAS: Columnas con tipo de dato número (float/int) que no sean binarias ni fechas
    list_numerical = [col for col in df.columns if pd.api.types.is_numeric_dtype(df[col])
                      and col not in list_binary and col not in list_dates]

    # 4. CATEGÓRICAS:
    list_categorical = [col for col in df.columns if col not in list_dates
                        and col not in list_binary and col not in list_numerical]

    return list_dates, list_binary, list_numerical, list_categorical

# Uso de la función
dates, bins, nums, cats = auto_classify_columns(df_reduced)

# Verificación rápida
print(f"Fechas: {dates}")
print(f"Binarias: {bins}")
print(f"Numéricas: {nums}")
print(f"Categorias: {cats}")

Fechas: ['pk_partition', 'entry_date']
Binarias: ['gender', 'active_customer', 'short_term_deposit', 'loans', 'mortgage', 'funds', 'securities', 'long_term_deposit', 'credit_card', 'payroll', 'pension_plan', 'payroll_account', 'emc_account', 'debit_card', 'em_account_p', 'em_acount']
Numéricas: ['pk_cid', 'age', 'salary', 'net_margin']
Categorias: ['region_code', 'entry_channel', 'segment']


In [ ]:
df_reduced[df_reduced['segment']=="Other"]['entry_channel'].value_counts()

,count
entry_channel,
Other,132423
KAT,439
KFC,438
KHM,160
KHN,141
KFA,84
RED,73
KHK,66
KHL,15


In [ ]:
df_reduced[df_reduced['segment']=="Other"]['emc_account'].value_counts()

,count
emc_account,
0,133832
1,112


Se observa que casi el 100% de los clientes que no tienen segmento definido (Tenian segmento nulo) no tienen el producto principal "em_account". Por tanto pueden tomar relevancia en el modelo como potenciales clientes para ofrecer el producto principal.

### DATA PREPARATION FECHAS

El tratamiento de las variables temporales se ha centrado en la descomposición de las fechas en componentes numéricos procesables por los algoritmos. Dada la naturaleza de los datos bancarios, se detectó una excepción crítica: la presencia de registros con etiquetas de texto como 'sin_venta' y valores nulos. Posteriormente, se extrajeron el año y el mes en columnas independientes, eliminando la columna original. Para garantizar que el modelo pudiera interpretar los casos sin fecha sin perder la fila, se imputaron estos valores con el valor constante -1.

In [ ]:
df_dates = df_reduced.copy()

In [ ]:
for col in dates:
    # Convertimos a datetime (esto detecta los 'sin_venta' y los hace NaT)
    temp_dt = pd.to_datetime(df_dates[col], errors='coerce')

    # Extraemos Año y Mes con tipos de datos pequeños
    df_dates[f'{col}_year'] = temp_dt.dt.year.fillna(-1).astype('int16')
    df_dates[f'{col}_month'] = temp_dt.dt.month.fillna(-1).astype('int8')

    # Eliminamos la columna original inmediatamente para liberar espacio
    df_dates.drop(columns=[col], inplace=True)

    # Liberamos memoria
    del temp_dt

# 3. Limpieza final de memoria
import gc
gc.collect()

# 4. Comprobación rápida (solo las nuevas columnas)
print("Columnas de fecha procesadas:")
print(df_dates.filter(like='_year').columns.tolist())
print(df_dates.filter(like='_month').columns.tolist())

Columnas de fecha procesadas:
['pk_partition_year', 'entry_date_year']
['pk_partition_month', 'entry_date_month']


In [ ]:
import gc

# 1. Eliminamos el dataset anterior.
if 'df_reduced' in locals():
    del df_reduced

# 2. Eliminamos listas o variables temporales pesadas
if 'df_merge' in locals():
    del df_merge

# 3. Forzamos la limpieza profunda
gc.collect()

0

### DATA PREPARATION NUMERICAS

Para las variables numéricas continuas, como el salario y la edad, el proceso se enfocó en mitigar el impacto de los valores atípicos (outliers) y simplificar la escala de los datos. Se identificó una excepción relevante en las distribuciones de ingresos y edades, con presencia de valores extremadamente altos y registros incoherentes como edades mínimas de 2 años,

Si bien estas edades podrían corresponder a cuentas de ahorro abiertas por tutores legales para el futuro de menores, se determinó que estos perfiles no son el público objetivo para el crecimiento de la cartera de productos activos.

Para corregirlo, se aplicó una técnica de capping, limitando los salarios al percentil 95 y restringiendo el rango de edad entre los 18 y 90 años. Una vez controlados los extremos, los valores faltantes se imputaron con la mediana de cada variable.

In [ ]:
df_num = df_dates.copy()

In [ ]:
for col in nums:
    # A. Conversión inicial
    temp_series = pd.to_numeric(df_num[col], errors='coerce')

    # B. Limpieza específica de OUTLIERS
    if 'salary' in col.lower():
        # Capamos salarios: todo lo que supere el percentil 95
        limite_superior = temp_series.quantile(0.95)
        temp_series = temp_series.clip(upper=limite_superior)

    elif 'age' in col.lower():
        # Capamos edades: entre 18 y 90 años
        # Los menores de 18 o mayores de 90 se ajustan a esos límites
        temp_series = temp_series.clip(lower=18, upper=90)

    temp_series = temp_series.round(2)

    # C. Imputación con la mediana del valor ya "limpio"
    mediana = temp_series.median()
    df_num[f'{col}_n'] = temp_series.fillna(mediana).astype('float32')

    # D. Eliminamos original
    df_num.drop(columns=[col], inplace=True)
    del temp_series

# 3. Limpieza de memoria
if 'df_dates' in locals():
    del df_dates
import gc
gc.collect()

30

### DATA PREPARATION CATEGORICAL

En cuanto a las variables cualitativas, el principal desafío fue la alta cardinalidad de ciertas dimensiones, lo que amenazaba con generar una matriz de datos excesivamente dispersa tras el proceso de codificación.

La estrategia consistió en una reducción de dimensionalidad selectiva: para cada variable, se preservaron únicamente los cinco niveles con mayor frecuencia de aparición.

Todos los valores restantes, considerados minoritarios o excepcionales, se reclasificaron bajo una etiqueta unificada denominada 'Otros'.

Una vez simplificadas, se transformaron mediante One-Hot Encoding para generar nuevas columnas binarias, eliminando sistemáticamente las variables originales.


In [ ]:
df_final = df_num.copy()

# Procesamos cada columna categórica
for col in cats:
    print(f"Transformando: {col}")

    # Identificamos los 5 valores más frecuentes
    top_7 = df_final[col].value_counts().nlargest(7).index.tolist()

    #  Creamos una versión temporal con los valores agrupados
    # Los que no son Top 5 pasan a ser 'Otros'
    temp_grouped = df_final[col].where(df_final[col].isin(top_7), 'Otros')

    # Generamos las columnas One-Hot Encoding (Dummies)
    # Usamos dtype='int8' para que no pese casi nada
    dummies = pd.get_dummies(temp_grouped, prefix=f"{col}_cat", drop_first=False, dtype='int8')

    # CONCATENAMOS las nuevas y ELIMINAMOS la vieja
    df_final = pd.concat([df_final, dummies], axis=1)
    df_final.drop(columns=[col], inplace=True)

    # Limpieza inmediata de RAM
    del temp_grouped
    del dummies
    gc.collect()

# 3. Borramos el dataset anterior
if 'df_num' in locals():
    del df_num

for ds in dfs:
    if ds in locals():
        del locals()[ds]



gc.collect()

Transformando: region_code
Transformando: entry_channel
Transformando: segment


0

In [ ]:
df_final.head()

,gender,active_customer,short_term_deposit,loans,mortgage,funds,securities,long_term_deposit,credit_card,payroll,...,entry_channel_cat_KHE,entry_channel_cat_KHK,entry_channel_cat_KHM,entry_channel_cat_KHQ,entry_channel_cat_Other,entry_channel_cat_Otros,segment_cat_01 - TOP,segment_cat_02 - PARTICULARES,segment_cat_03 - UNIVERSITARIO,segment_cat_Other
0,H,1.0,0,0,0,0,0,0,0,0.0,...,0,0,0,0,0,1,0,1,0,0
1,V,0.0,0,0,0,0,0,0,0,0.0,...,1,0,0,0,0,0,0,0,1,0
2,V,0.0,0,0,0,0,0,0,0,0.0,...,1,0,0,0,0,0,0,0,1,0
3,H,0.0,1,0,0,0,0,0,0,0.0,...,0,0,0,0,0,1,0,0,1,0
4,V,1.0,0,0,0,0,0,0,0,0.0,...,1,0,0,0,0,0,0,0,1,0


In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5961849 entries, 0 to 5962923
Data columns (total 44 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   gender                          object 
 1   active_customer                 float64
 2   short_term_deposit              int64  
 3   loans                           int64  
 4   mortgage                        int64  
 5   funds                           int64  
 6   securities                      int64  
 7   long_term_deposit               int64  
 8   credit_card                     int64  
 9   payroll                         float64
 10  pension_plan                    float64
 11  payroll_account                 int64  
 12  emc_account                     int64  
 13  debit_card                      int64  
 14  em_account_p                    int64  
 15  em_acount                       int64  
 16  pk_partition_year               int16  
 17  pk_partition_month              

### Data Preparation Boolean

La fase final de la preparación consistió en la homogeneización de todas las variables binarias y de estado al tipo de dato lógico boolean.

En este bloque se gestionó la excepción de la variable gender, que originalmente se presentaba como una cadena de texto; esta se mapeó a valores numerico (1/0) para integrarla eficientemente con el resto del dataset.

De igual manera, las variables de tenencia de productos se transformaron desde su formato numérico a booleano, interpretando cualquier valor nulo como una ausencia de contratación.


In [ ]:
if 'gender' in df_final.columns:
  '''
  Tal y como se observo en el data understanding, el atributo Gender es un atributo
  en formato de string, po lo que es necesario convertirlo en un formato numerico
  o booleano

  '''
df_final['gender_b'] = df_final['gender'].map({'V': 1, 'H': 0}).fillna(False).astype('int8')

# Borramos la original
df_final.drop(columns=['gender'], inplace=True)

# 2. Procesamos el resto de las binarias
for col in bins:
    if col in df_final.columns: # Por si acaso alguna ya se borró
        # Creamos la nueva booleana
        df_final[f'{col}_b'] = pd.to_numeric(df_final[col], errors='coerce').fillna(0).astype('int8')

        # Eliminamos la original
        df_final.drop(columns=[col], inplace=True)

gc.collect()

0

Finalmente, una vez realizado el Data Preparation, el dataset resultante se presenta como una estructura optimizada y libre de ruido. Se ha logrado transformar un conjunto de datos bruto en un activo analítico donde las fechas son componentes numéricos, las variables continuas han sido saneadas de valores atípicos, y las categorías se han simplificado para evitar una dimensionalidad excesiva.

In [ ]:
df_final.head()

,pk_partition_year,pk_partition_month,entry_date_year,entry_date_month,pk_cid_n,age_n,salary_n,net_margin_n,region_code_cat_8.0,region_code_cat_15.0,...,securities_b,long_term_deposit_b,credit_card_b,payroll_b,pension_plan_b,payroll_account_b,emc_account_b,debit_card_b,em_account_p_b,em_acount_b
0,2018,1,2018,1,1375586.0,35.0,87218.101562,0.0,0,0,...,0,0,0,0,0,0,0,0,0,1
1,2018,1,2015,8,1050611.0,23.0,35548.738281,0.0,0,0,...,0,0,0,0,0,0,0,0,0,1
2,2018,1,2015,8,1050612.0,23.0,122179.109375,0.0,0,0,...,0,0,0,0,0,0,0,0,0,1
3,2018,1,2015,8,1050613.0,22.0,119775.539062,0.0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,2018,1,2015,8,1050614.0,23.0,88441.468750,0.0,0,0,...,0,0,0,0,0,0,0,0,0,1


In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5961849 entries, 0 to 5962923
Data columns (total 44 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   pk_partition_year               int16  
 1   pk_partition_month              int8   
 2   entry_date_year                 int16  
 3   entry_date_month                int8   
 4   pk_cid_n                        float32
 5   age_n                           float32
 6   salary_n                        float32
 7   net_margin_n                    float32
 8   region_code_cat_8.0             int8   
 9   region_code_cat_15.0            int8   
 10  region_code_cat_28.0            int8   
 11  region_code_cat_29.0            int8   
 12  region_code_cat_30.0            int8   
 13  region_code_cat_41.0            int8   
 14  region_code_cat_46.0            int8   
 15  region_code_cat_Otros           int8   
 16  entry_channel_cat_KAT           int8   
 17  entry_channel_cat_KFC           

In [ ]:
ruta_guardado = '/content/Drive/MyDrive/datasets_TFM/df_final_preparado.parquet'
df_final.to_parquet(ruta_guardado)